### ISSUE TITLE
workspace CLI: mets_basename is ignored

### ISSUE THREAD

**Issue by bertsky** [OCR-D/core#517]
`ocrd workspace` announces itself with the following options to control where the workspace's local METS is located:
```
 -d, --directory WORKSPACE_DIR  Changes the workspace folder location.
                                 [default: .]
  -M, --mets-basename TEXT       The basename of the METS file.  [default:
                                 mets.xml]
```

So when it is not under CWD or not called `mets.xml`, I must use these options.

First, allow me to point out that this is in contrast to all the processor CLIs which have but a single (and more memorable) option for both at the same time:
```
  -m, --mets URL-PATH             URL or file path of METS to process
```

However, the workspace CLI does not always _implement_ the `--mets-basename` option correctly:
- some commands respect it: `init`, `clone`, `add-file`, `bulk-add`, `remove-file`, `remove-group`, `prune-files` and `set-id`
- other commands ignore it: `validate`, `list-group`, `list-page` and `get-id`

So, when my METS uses a different basename, the latter commands will just fail like this:
```
Exception: File does not exist: mets.xml
```

This can easily be fixed of course. 

But how about _also_ adding a `-m, --mets` option as an alternative? (The implementation would derive the directory and basename from it, but complain when both options are used inconsistently.)

**Comment by kba** [comment]
`mets_basename` was a bad idea from the beginning and I'd be happy to be rid of it. So, I'd rather just replace `-M/--mets-basename` with `-m/--mets throughout. 

>  but complain when both options are used inconsistently

In other words:

  * If only `--directory` is set and `--mets` is not: Assume `--mets` to be `--directory/mets.xml`
  * If only `--mets` is set and a filename: If `--mets` has a dirname, set `--dirname` to `dirname(--mets`)
  * If both `--directory` and `--mets` are set: If `--mets` has a dirname: validate that `--dirname == dirname(--mets)`. Otherwise, prepend `--dirname` to `--mets`.

This way `--mets` can function like `--mets-basename` depending on `--directory`, right?

Should of course be tested but since this doesn't touch on processors, it wouldn't require a spec change and reduce complexity, so :+1:.

**Comment by bertsky** [comment]
> `mets_basename` was a bad idea from the beginning and I'd be happy to be rid of it. So, I'd rather just replace `-M/--mets-basename` with `-m/--mets` throughout.

I see.

(So as with the processor CLIs, `--directory` would always be strictly local, whereas `--mets` could also be remote, in which case the METS gets cloned into `directory` first – right?)

> In other words:
> 
>     * If only `--directory` is set and `--mets` is not: Assume `--mets` to be `--directory/mets.xml`
> 
>     * If only `--mets` is set and a filename: If `--mets` has a dirname, set `--dirname` to `dirname(--mets`)
> 
>     * If both `--directory` and `--mets` are set: If `--mets` has a dirname: validate that `--dirname == dirname(--mets)`. Otherwise, prepend `--dirname` to `--mets`.
> 
> 
> This way `--mets` can function like `--mets-basename` depending on `--directory`, right?

Yes, exactly – I would say so!

> Should of course be tested but since this doesn't touch on processors, it wouldn't require a spec change and reduce complexity, so +1.

Not sure whether this couldn't break some existing module tests, though.

**Comment by kba** [comment]
> (So as with the processor CLIs, --directory would always be strictly local, whereas --mets could also be remote, in which case the METS gets cloned into directory first – right?)

`ocrd workspace clone` is an outlier here because `METS_URL` is also the first CLI argument. I think the same logic w.r.t. `--mets`/`--filename` should apply. I can't right now think of an example where `--mets` might be a remote URL.

**Comment by kba** [comment]
And I'm doubting implicit cloning in general these days. While a neat and fun feature and arguably really practical in some cases, it requires a lot of if-else and makes the code harder to maintain. I think it's easier in the long run to only allow that for `clone` and maybe `validate` but nowhere else.

**Comment by bertsky** [comment]
> > (So as with the processor CLIs, --directory would always be strictly local, whereas --mets could also be remote, in which case the METS gets cloned into directory first – right?)
> 
> `ocrd workspace clone` is an outlier here because `METS_URL` is also the first CLI argument.

It is – but I was not referring to cloning as a workspace command. I was referring to the implicit _download_ of the `--mets` into `--directory` if remote. This already works for all processor CLIs!

>  I think the same logic w.r.t. `--mets`/`--filename` should apply. I can't right now think of an example where `--mets` might be a remote URL.

For `ocrd workspace clone` yes, it does not make sense for `--mets` to be remote itself (i.e. cloning from remote in order to clone again locally).

But the clone command's current interface has a catch itself: because it _also_ has the optional `WORKSPACE_DIR` argument, the `--directory` option just gets ignored. (I believe the last argument should be dropped/deprecated.)

> And I'm doubting implicit cloning in general these days. While a neat and fun feature and arguably really practical in some cases, it requires a lot of if-else and makes the code harder to maintain. 

But what about the upcoming generalization from CLI to REST API? Processors will (have to) get called with METS URLs at some point, won't they?

> I think it's easier in the long run to only allow that for `clone` and maybe `validate` but nowhere else.

The `validate` command's current implementation is even more problematic, as it…
- ignores `mets-basename` completely (only tries `mets.xml`)
- has n-ary `METS_URL`, but ignores all but the first

If we were to adopt the `--directory --mets` scheme for all workspace CLI commands, then `validate` would have to get rid of its `METS_URL` argument completely IMO.

**Comment by kba** [comment]
Fixed in [v2.15.0](https://github.com/OCR-D/core/releases/v2.15.0)
